### PR TITLE
Fix #433 bug where ListConfig is modified after assignment of wrong value.

### LINKED ISSUE #433: Assignment of invalid value to ListConfig modifies the value.
**Describe the bug**
Assigning a list with invalid values to a ListConfig keeps the content modified after raising the error.
**To Reproduce**
```python
    @dataclass
    class ListClass:
        list: List[int] = field(default_factory=lambda: [1])
    cfg = OmegaConf.structured(ListClass)
    try:
        cfg.list = [1, 2, "foo"]
    except:
        pass
    print(cfg.list) # output [1, 2]

```

**Expected behavior**
cfg.list should return to be [1] 
**Additional context**
- [x] OmegaConf version: 2.1.0
- [x] Python version: 3.8
- [x] Operating system : macOS
- [x] Please provide a minimal repro

### PR DESCRIPTION
**Pull request by pereman2**
Closes #433 .

### CODE REVIEW

File: news/433.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix assignment of an invalid value to a ListConfig to raise an exception without modifying the config object
```
Comment (omry) [inline news/433.bugfix:1]: Describe the bug this is fixing, not the fix.
```suggestion
Fix bug where assignment of an invalid value to a ListConfig raised an exception but left the object modified.
```

File: omegaconf/listconfig.py [hunk 2]
```diff
@@ -504,6 +504,16 @@ def __contains__(self, item: Any) -> bool:
         return False
 
     def _set_value(self, value: Any, flags: Optional[Dict[str, bool]] = None) -> None:
+        try:
+            previous_content = self.__dict__["_content"]
```
Comment (omry) [inline omegaconf/listconfig.py:508]: interesting that there is no need for a shallow copy here (unlike the DictConfig). how come?

**Review by omry — CHANGES_REQUESTED** [review]
Nice. a couple off comments.

File: omegaconf/listconfig.py (hunk 2 above)
Comment (pereman2) [reply omegaconf/listconfig.py:508]: `self.__dict__["_content"] = {}`allocates a new pointer thus making `previous_content` able to be assigned the last content state without needing a copy. I forgot to ask you if I should copy this behaviour in dictconfig in this same diff.

Comment (omry) [reply omegaconf/listconfig.py:508]: Not sure I follow what behavior.
maybe separate PR to clean DictConfig's shallow copy?

**Review by omry — APPROVED** [review]

File: omegaconf/listconfig.py (hunk 2 above)
Comment (pereman2) [reply omegaconf/listconfig.py:508]: Okay

**Pull request merged**